npm-cli/src/Pesy.re

```diff
@@ -36,188 +24,99 @@ let substituteTemplateValues = s =>
-/*       packageNameUpperCamelCase, */
-/*     ), */
-/*     spf( */
-/*       "require: [%s]", */
-/*       List.fold_left( */
-/*         (acc, e) => acc ++ " " ++ e, */
-/*         "", */
-/*         [spf("\"%s\"", packageLibName)], */
-/*       ), */
-/*     ), */
-/*   ], */
-/*   ["library", "require: []"], */
-/*   [ */
-/*     "test", */
-/*     spf( */
-/*       "bin:    Test%s.re as Test%s.exe", */
-/*       packageNameUpperCamelCase, */
-/*       packageNameUpperCamelCase, */
-/*     ), */
-/*     spf( */
-/*       "require: [%s]", */
-/*       String.concat(", ", [spf("\"%s\"", packageLibName)]), */
-/*     ), */
-/*   ], */
-/* ]); */
-/* print_newline(); */
-
-let id = Spinner.start("\x1b[2mRunning\x1b[0m esy install");
-exec("esy i", (e, stdout, stderr) => {
-  Spinner.stop(id);
-  if (Js.eqNullable(e, Js.Nullable.null)) {
-    let id = Spinner.start("\x1b[2mRunning\x1b[0m esy pesy\x1b[2m and \x1b[0m building project dependencies");
-    exec("esy pesy", (e, stdout, stderr) => {
-      Spinner.stop(id);
-      if (Js.eqNullable(e, Js.Nullable.null)) {
-        let id = Spinner.start("\x1b[2mRunning\x1b[0m esy build");
-        exec("esy build", (e, stdout, stderr) => {
-          Spinner.stop(id);
-          if (Js.eqNullable(e, Js.Nullable.null)) {
-            Printf.printf("You may now run \x1b[32m'esy test'\x1b[0m\n");
-          } else {
-            Printf.printf(
-              "'esy build' \x1b[31mfailed.\x1b[0m Could not build project.\nLogs can be found in pesy.stdout.log and pesy.stderr.log\n",
-            );
-            write("pesy.stdout.log", stdout);
-            write("pesy.stderr.log", stderr);
-          };
-        });
-      } else {
-        Printf.printf(
-          "'esy pesy' \x1b[31mfailed.\x1b[0m Dune files could not be created.\n Logs can be found in pesy.stdout.log and pesy.stderr.log\n",
+// TODO: Move the template to the esy or esy-ocaml org
+let template =
+  argv
+  ->Belt.Array.keep(Js.String.includes("--template"))
+  ->Belt.Array.get(0)
+  ->Belt.Option.map(Js.String.replace("--template=", ""))
+  ->Belt.Option.getWithDefault("github:ulrikstrid/hello-reason-pesy");
+
+let download_spinner =
+  Spinner.start("\x1b[2mDownloading template\x1b[0m " ++ template);
+download_git(
+  template,
+  projectPath,
+  error => {
+    Spinner.stop(download_spinner);
+    switch (error) {
+    | Some(e) => Js.log(e)
+    | None =>
+      let setup_files_spinner =
+        Spinner.start("\x1b[2mSetting up template\x1b[0m " ++ template);
+      let files =
+        walk_sync(projectPath)
+        ->Belt.Array.keep(file_or_dir => statSync(file_or_dir)->isFile);
+
+      Belt.Array.forEach(
+        files,
+        file => {
+          let () = readFile(file)->substituteTemplateValues |> write(file);
+
+          renameSync(file, substituteTemplateValues(file));
+        },
+      );
+
+      let libKebab = packageNameKebabSansScope;
+      let duneProjectFile = Path.(projectPath / "dune-project");
```

Comment:
Isn't `dune-project` file a part of the hello-reason-pesy repo?

Reply:
It is, but in this case we're generating a `dune-project` and `package-name.opam`if the template doesn't include them for some reason.
But I could just remove these, in that case we could potentially support bucklescript as well if it works with `esy` in the future.

Reply:
Well, then this brings up the issue of validation of the templates. For now let's track this in a separate issue and assume the template has all the files needed.

Reply:
https://github.com/esy/pesy/issues/38